hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMTrashRecoverRequest.java

```diff
@@ -0,0 +1,138 @@
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.om.request.key;
+
+import java.io.IOException;
+
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.ozone.om.ratis.utils.OzoneManagerDoubleBufferHelper;
+import org.apache.hadoop.ozone.om.response.key.OMTrashRecoverResponse;
+import org.apache.hadoop.ozone.security.acl.IAccessAuthorizer;
+import org.apache.hadoop.ozone.security.acl.OzoneObj;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.ozone.om.OMMetadataManager;
+import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .RecoverTrashRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Type;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Status;
+
+import static org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource.BUCKET_LOCK;
+
+/**
+ * Handles RecoverTrash request.
+ */
+public class OMTrashRecoverRequest extends OMKeyRequest {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMTrashRecoverRequest.class);
+
+  public OMTrashRecoverRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) {
+    RecoverTrashRequest recoverTrashRequest = getOmRequest()
+        .getRecoverTrashRequest();
+    Preconditions.checkNotNull(recoverTrashRequest);
+
+    return getOmRequest().toBuilder().build();
+  }
+
+  @Override
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long transactionLogIndex,
+      OzoneManagerDoubleBufferHelper ozoneManagerDoubleBufferHelper) {
+    RecoverTrashRequest recoverTrashRequest = getOmRequest()
+        .getRecoverTrashRequest();
+    Preconditions.checkNotNull(recoverTrashRequest);
+
+    String volumeName = recoverTrashRequest.getVolumeName();
+    String bucketName = recoverTrashRequest.getBucketName();
+    String keyName = recoverTrashRequest.getKeyName();
+    String destinationBucket = recoverTrashRequest.getDestinationBucket();
+
+    /** TODO: HDDS-2818. New Metrics for Trash Key Recover and Fails.
+     *  OMMetrics omMetrics = ozoneManager.getMetrics();
+     */
+
+    OMResponse.Builder omResponse = OMResponse.newBuilder()
+        .setCmdType(Type.RecoverTrash).setStatus(Status.OK)
+        .setSuccess(true);
+
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    boolean acquireLock = false;
+    OMClientResponse omClientResponse = null;
+    try {
+
+      // check Acl
+      checkKeyAcls(ozoneManager, volumeName, destinationBucket, keyName,
```

Comment:
For checkAcl, the key does not exist, checkKeyAcls will fail with KeyNotFound.
 So, I think we should check Bucket write Acls here?
cc @anuengineer

Reply:
Yes, thanks for this comment.
I think we should check bucket here. (For the deleted key not found.)
Fixed.